specs/WebMessageObjects.md

```diff
@@ -0,0 +1,355 @@
+
+            for (unsigned int i = 0; i < length; i++)
+            {
+                wil::com_ptr<IUnknown> object;
+                objectsCollection->GetValueAtIndex(i, &object);
+
+                wil::com_ptr<ICoreWebView2File> file =
+                    object.query<ICoreWebView2File>();
+                if (file)
+                {
+                    // Add the file to message to be sent back to webview
+                    wil::unique_cotaskmem_string path;
+                    file->get_Path(&path);
+                    paths.push_back(path.get());
+                }
+            }
+
+            ProcessPaths(paths);
+            return S_OK;
+        })
+        .Get(),
+    &m_webMessageReceivedToken);
+```
+
+```c#
+webView.CoreWebView2.WebMessageReceived += WebView_WebMessageReceivedHandler;
+void WebView_WebMessageReceivedHandler(object sender, CoreWebView2WebMessageReceivedEventArgs args)
+{
+    List<string> paths = new List<string>();
+    foreach (var object in args.AdditionalObjects)
+    {
+        if (object is CoreWebView2File)
+        {
+            paths.Add(((CoreWebView2File)object).Path);
+        }
+    }
+    ProcessPaths(paths);
+}
+```
+
+# API Details
+## SDK API
+```c#
+/// Representation of a DOM
+/// (File)[https://developer.mozilla.org/en-US/docs/Web/API/File] object
+/// passed via WebMessage. You can use this object to obtain the path of a
+/// File dropped on WebView2.
+/// \snippet ScenarioDragDrop.cpp DroppedFilePath
+[uuid(f2c19559-6bc1-4583-a757-90021be9afec), object, pointer_default(unique)]
+interface ICoreWebView2File : IUnknown {
+  /// Get the absolute file path.
+  [propget] HRESULT Path([out, retval] LPWSTR* path);
+}
+
+/// Read-only collection of generic objects
+[uuid(b547d2d4-b8d5-4278-a544-c54e66b5f45d), object, pointer_default(unique)]
+interface ICoreWebView2ObjectCollectionView : IUnknown {
+  /// Gets the number of items in the collection.
+  [propget] HRESULT Count([out, retval] UINT32* value);
+
+  /// Gets the object at the specified index. Cast the object to the native type
+  /// to access its specific properties.
+  HRESULT GetValueAtIndex([in] UINT32 index,
+      [out, retval] IUnknown** value);
+}
+
+[uuid(50a798ac-40fa-4634-8fb7-9419e5e8a3f8), object, pointer_default(unique)]
+interface ICoreWebView2WebMessageReceivedEventArgs2 : ICoreWebView2WebMessageReceivedEventArgs {
+  /// Additional received WebMessage objects. To pass `additionalObjects` via
+  /// WebMessage to the app, use the
+  /// `chrome.webview.postMessageWithAdditionalObjects` content API.
+  /// Any DOM object type that can be natively representable that has been
+  /// passed in to `additionalObjects` parameter will be accessible here.
+  /// Currently a WebMessage object can be the following type:
+  /// - `ICoreWebView2File`.
+  [propget] HRESULT AdditionalObjects(
+      [out, retval] ICoreWebView2ObjectCollectionView** value);
+}
+```
+
+```c#
+namespace Microsoft.Web.WebView2.Core

```

Comment:
Yes, because we propose to type map the native ICoreWebView2File to System.IO.FileInfo in .NET and Windows.Storage.StorageFile in WinRT. ObjectCollectionView will simply be exposed as a `IVectorView<object>.

Reply:
Yes. No change.